sprint_9-solution-http-api

A-Glebov opened the pull request:
1. Реализовали API, в котором эндпоинты соответствуют вызовам базовых методов интерфейса TaskManager
2. Добавили базовый класс BaseHttpHandler общий для всех обработчиков
3. Написали обработчики для всех базовых путей
4. Написали тесты для тестирования нового функционала

A-Glebov committed:
1. Реализовали API, в котором эндпоинты соответствуют вызовам базовых методов интерфейса TaskManager
2. Добавили базовый класс BaseHttpHandler общий для всех обработчиков
3. Написали обработчики для всех базовых путей
4. Написали тесты для тестирования нового функционала

Ivanov-s-tmn commented on java-kanban.iml:
директория .idea и файл .iml генерируются локально ide и не нужны в репозитории, предлагаю добавить их в исключения gitignore

Ivanov-s-tmn commented on src/adapter/DurationTypeAdapter.java:
```diff
@@ -0,0 +1,54 @@
+        // Преобразуем Duration в строку формата ISO-8601
+        return context.serialize(duration.toString());
+    }
+}*/
```
Лишний код стараемся сразу удалять и не засорять проект

Ivanov-s-tmn commented on src/http/EpicHandler.java:
```diff
@@ -0,0 +1,104 @@
+                                sendNotFound(exchange, "Эпик с ID=" + id + " не найден");
+                                return;
+                            }
+                            String response = gson.toJson(epic.getSubTaskIdList());
```
Тут нужно вернуть список подзадач эпика, для этого есть метод `getListOfSubtaskOfEpic` в таск менеджере

Ivanov-s-tmn commented on src/http/EpicHandler.java:
```diff
@@ -0,0 +1,104 @@
+                        int id = parseId(pathId);
+
+                        if (id != -1) {
+                            Epic epic = taskManager.getEpic(id);
```
Лишний раз вызывать методы таск менеджера не нужно, тут к примеру, сначала обращаемся к таск менеджеру за эпиком, но у этого метода есть дополнительная логика по добавлению в историю, а нам это не нужно в данном случае

Ivanov-s-tmn replied:
Этот вызов, как я и писал, не нужен. Метод `getListOfSubtaskOfEpic `должен работать по айди эпика, а не по самому эпику, тогда отпадет необходимость в лишнем вызове метода `getEpic`
`taskManager.getListOfSubtaskOfEpic(id);`

Ivanov-s-tmn commented on src/http/EpicHandler.java:
```diff
@@ -0,0 +1,104 @@
+
+                }
+
+
```
Лишние отступы не оставляем, не более одной пустой строки между частями кода

Ivanov-s-tmn commented on src/http/EpicHandler.java:
```diff
@@ -0,0 +1,104 @@
+                        String body = new String(exchange.getRequestBody().readAllBytes(), StandardCharsets.UTF_8);
+                        Epic epic = gson.fromJson(body, Epic.class);
+                        taskManager.createEpic(epic);
+
+                        if (taskManager.getListOfEpics().contains(epic)) {
+                            sendText(exchange, "Эпик успешно добавлен");
+                        }
+
```
Такая проверка не должна быть частью логики хендлера, факт добавления в менеджер мы уже проверим в тестах. А логика добавления эпика должна уже быть полностью реализована в самом таск менеджере

Метод создания ничего не возвращает и в ответ просто возвращаем статус `exchange.sendResponseHeaders(200, 0);`

Ivanov-s-tmn commented on src/http/EpicHandler.java:
```diff
@@ -0,0 +1,104 @@
+                        String pathId = path.replaceFirst("/epics/", "");
+                        int id = parseId(pathId);
+
+                        if (id != -1 && taskManager.getEpic(id) != null) {
```
Тут тоже лишний вызов `getEpic`

Ivanov-s-tmn commented on src/http/HttpTaskServer.java:
```diff
@@ -0,0 +1,74 @@
+    public static final int PORT = 8080;
+    private static final String URL = "http://localhost:";
+    private final HttpServer httpServer;
+    TaskManager taskManager;
```
Стоит соблюдать инкапсуляцию и делать поля как можно более закрытыми

Ivanov-s-tmn commented on src/http/SubTaskHandler.java:
К хендлеру сабтасок и тасок аналогичные замечания, которые я написал к хендлеру эпиков

Ivanov-s-tmn commented on src/service/InMemoryTaskManager.java:
Краткий фидбек по таск менеджеру:
1. Методы не используются из вне и вызываются только в логике других методов менеджера `checkTaskForIntersectionInTime`, `validateTaskByDateTime`, `getTaskId()`, `updateEpicStatus` не должны быть в публичном доступе из интерфейса `TaskManager` и их стоит сделать приватными. Плюс, обычно приватные методы размещают после публичных для облегчения чтения кода сверху вниз
2. В методах `deleteAllTasks()` и `deleteAllSubTask()` отсортированный список prioritizedTasks не должен очищаться полностью, должны удаляться только задачи (`deleteAllTasks()`) и подзадачи (`deleteAllSubTask()`)
В методе `deleteAllEpic()` должны удаляться и сабтаски + из отсортированного списка. Сабтаски без эпика существовать не могут
3. В методах `getTask()` и аналогичных для сабтасок и эпиков не лучшая практика возвращать null, `return tasks.get(id);` и так вернет null или объект, если он есть в хешмапе. А если эта проверка сделана для вызова добавления в историю, то достаточно простой проверки на null в методе `add()` менеджера истории
4. Проверять на пересечение по времени нужно не только при создании, но и при обновлении

Ivanov-s-tmn commented on src/service/Managers.java:
```diff
@@ -1,7 +1,15 @@
 package service;
 
+//import adapter.DurationSerializer;
```
Лишнее

Ivanov-s-tmn commented on test/http/HttpTaskManagerEpicsTest.java:
```diff
@@ -0,0 +1,159 @@
+        subtask = new SubTask(2, Type.SUBTASK, "Subtask Epic", "Description Sub Ep",
+                Status.NEW, LocalDateTime.of(2025, 1, 1, 1, 0),
+                Duration.ofMinutes(30), 1);
+
+        HttpRequest request1 = HttpRequest.newBuilder().uri(URI.create("http://localhost:8080/epics"))
+                .POST(HttpRequest.BodyPublishers.ofString(gson.toJson(epic))).build();
+        HttpResponse<String> response1 = client.send(request1, HttpResponse.BodyHandlers.ofString());
+
+        HttpRequest request2 = HttpRequest.newBuilder().uri(URI.create("http://localhost:8080/subtasks"))
+                .POST(HttpRequest.BodyPublishers.ofString(gson.toJson(subtask))).build();
+        HttpResponse<String> response2 = client.send(request2, HttpResponse.BodyHandlers.ofString());
```
Перед тестами не нужно делать запрос через http для создания, создать через менеджер, а в тестах где нужно проверить удаление/получение отправить запрос через http DELETE/GET и проверить результат

A-Glebov committed:
1. В EpicHandler изменили возвращаемы список с Id подзадач на список подзадач.
2. В обработчиках заменили методы влияющие на приоритет задач
3. checkTaskForIntersectionInTime, validateTaskByDateTime, updateEpicStatus сделали приватными
4. Исправили в методах deleteAllTasks() и deleteAllSubTask() отсортированный список prioritizedTasks теперь не очищается полностью
5. Добавили проверку на пересечение по времени не только при создании, но и при обновлении
6. Задачи для тестирования добавили через менеджер, а не через http запрос

Ivanov-s-tmn commented on src/http/EpicHandler.java:
```diff
@@ -27,12 +27,12 @@ public void handle(HttpExchange exchange) {
 
                         if (id != -1) {
-                            Epic epic = taskManager.getEpic(id);
+                            Epic epic = taskManager.getEpics().get(id);
```
Это тоже не выход, сейчас через метод `getEpics()` можно получить ссылку на исходную хешмапу в памяти и к объектам в памяти, что небезопасно с точки зрения разработки. Зачем в этом случае вообще проверка, есть этот эпик или нет? Если эпика нет, просто должен вернуться пустой список и это все должно быть уже учтено в логике метода `getListOfSubtaskOfEpic()`

Ivanov-s-tmn commented on src/http/EpicHandler.java:
```diff
@@ -82,7 +81,7 @@ public void handle(HttpExchange exchange) {
                         int id = parseId(pathId);
 
-                        if (id != -1 && taskManager.getEpic(id) != null) {
+                        if (id != -1 && taskManager.getEpics().containsKey(id)) {
```
`taskManager.getEpics().containsKey(id)` тоже излишне, удаление не должно вызывать ошибок, если мы попытаемся удалить эпик которого не существует, то в этом нет ничего страшного и ошибок возникать не должно

Ivanov-s-tmn commented on src/http/SubTaskHandler.java:
```diff
@@ -67,7 +68,7 @@ public void handle(HttpExchange exchange) {
                             SubTask subtask = gson.fromJson(body, SubTask.class);
 
-                            if (taskManager.getSubTask(id) != null) {
+                            if (taskManager.getSubtasks().containsKey(id)) {
```
Эти проверки аналогично нужно убрать

Ivanov-s-tmn commented on src/http/TaskHandler.java:
```diff
@@ -60,15 +61,15 @@ public void handle(HttpExchange exchange) {
                             Task task = gson.fromJson(body, Task.class);
 
-                            if (taskManager.getTask(id) != null) {
+                            if (taskManager.getTasks().containsKey(id)) {
```
И эти

Ivanov-s-tmn commented on src/service/InMemoryTaskManager.java:
```diff
@@ -17,72 +17,39 @@ public class InMemoryTaskManager implements TaskManager {
-        return (task.getStartTime().isAfter(existTask.getStartTime()) && task.getStartTime().isBefore(existTask.getEndTime()))
-                || (task.getEndTime().isAfter(existTask.getStartTime()) && task.getEndTime().isBefore(existTask.getEndTime()))
-                || task.getStartTime().equals(existTask.getStartTime())
-                || task.getEndTime().equals(existTask.getEndTime());
+    @Override
+    public HashMap<Integer, Epic> getEpics() {
+        return epics;
     }
 
     @Override
-    public boolean validateTaskByDateTime(Task task) {
-        return prioritizedTasks.stream()
-                .anyMatch(existTask -> checkTaskForIntersectionInTime(task, existTask));
+    public HashMap<Integer, SubTask> getSubtasks() {
+        return subtasks;
```
Эти методы нужно убрать

Ivanov-s-tmn commented on src/service/InMemoryTaskManager.java:
```diff
@@ -17,72 +17,39 @@ public class InMemoryTaskManager implements TaskManager {
-        epic.setDuration(duration);
-
+    @Override
+    public int getTaskId() {
+        ++taskId;
+        return taskId;
```
Этот метод не должен быть в открытом доступе, нужно генерировать и присваивать задачам айди перед добавлением в хешмапы в методах create

A-Glebov committed:
1. Убрали лишние проверки из обработчиков
2. Лишние методы из TaskManager
3. Сделали getTaskId() private. Добавили его вызов при создании задачи
4. В BaseHttpHandler добавили метод sendCreated() для отправки кода 201 так же добавили его вызов в логику обработчиков
5. В InMemoryManager в методы обновления задач добавили проверку на null для исключения добавления задачи если она отсутствует в менеджере
